controller/lib/hal/stepper.cpp

```diff
@@ -0,0 +1,820 @@
+  if (speed > 0x3ff)
+    speed = 0x3ff;
+
+  return SetParam(StepMtrParam::MAX_SPEED, speed);
+}
+
+// Get the motor's max speed setting in deg/sec
+StepMtrErr StepMotor::GetMaxSpeed(float *ret) {
+  uint32_t val;
+  StepMtrErr err = GetParam(StepMtrParam::MAX_SPEED, &val);
+  *ret = RegVelToDps(val, cvt_vel_max_speed_reg);
+  return err;
+}
+
+// Set the motor's min speed setting in deg/sec
+//
+// NOTE - Setting a non-zero minimum speed doesn't mean
+// the motor can't stop, this is the minimum speed for
+// a move.  When you start a move, rathern then increase
+// linearly from 0, the stepper will jump to this speed
+// immediately, then ramp up from there.
+// This can help with vibration.
+//
+// NOTE - The motor must be disabled to set this
+StepMtrErr StepMotor::SetMinSpeed(float dps) {
+  if (dps < 0)
+    return StepMtrErr::BAD_VALUE;
+
+  uint32_t speed =
+      static_cast<uint32_t>(DpsToVelReg(dps, cvt_vel_min_speed_reg));
+  if (speed > 0xfff)
+    speed = 0xfff;
+
+  return SetParam(StepMtrParam::MIN_SPEED, speed);
+}
+
+// Get the motor's min speed setting in deg/sec
+StepMtrErr StepMotor::GetMinSpeed(float *ret) {
+  uint32_t val;
+  StepMtrErr err = GetParam(StepMtrParam::MIN_SPEED, &val);
+  *ret = RegVelToDps(val, cvt_vel_min_speed_reg);
+  return err;
+}
+
+// Set the motors accel and decel rate in deg/sec/sec units
+//
+// NOTE - The motor must be disabled to set this
+StepMtrErr StepMotor::SetAccel(float acc) {
+
+  static const float scaler = (tick_time * tick_time * 1099511627776.f);
+  if (acc < 0)
+    return StepMtrErr::BAD_VALUE;
+
+  // Convert from deg/sec/sec to steps/sec/sec
+  acc *= static_cast<float>(steps_per_rev_) / 360.0f;
+
+  // Convert to the proper units for the driver chip
+  uint32_t val = static_cast<uint32_t>(acc * scaler);
+  if (val > 0x0fff)
+    val = 0xfff;
+
+  StepMtrErr err = SetParam(StepMtrParam::ACCEL, val);
+  if (err != StepMtrErr::OK)
+    return err;
+
+  return SetParam(StepMtrParam::DECEL, val);
+}
+
+// Set the amplitude of the voltage output used to drive the motor.
+// The values set here allow the amplitude of output that pushes power
+// to the motor to be adjusted.  Higher values will push more current
+// into the motor, lower values will push less.  The motor will be
+// more powerful with higher values, but will get hotter, consume more
+// power, and could potentially be damaged if these are set too high,
+// so be careful.
+//
+// There are four different values that can be set which control the
+// output to the motor in different phases of it's motion:
+//   hold - Value used when the motor is holding position (not moving)
+//   run  - Value used when running at constant velocity.
+//   accel - Value used when accelerating
+//   decel - Value used when decelerating
+//
+// In all cases the values are set in a range of 0 to 1
+// for 0 to 100%
+StepMtrErr StepMotor::SetKval(StepMtrParam param, float amp) {
+
+  if ((amp < 0) || (amp > 1))
```

Comment:
parens :)

You may have been bitten by order-of-operations bugs, but compilers are pretty good at catching those nowadays.  e.g. `x & y == 0` is a warning, yay.

Reply:
Yup, have many scars!